nemo/collections/llm/peft/lora.py

```diff
@@ -54,46 +66,117 @@ def __init__(
         lora_A_init_method='xavier',
         lora_dtype=None,
     ):
-        super(LinearAdapter, self).__init__()
         assert isinstance(orig_linear, nn.Linear)
+        super(LinearAdapter, self).__init__(
+            in_features=orig_linear.in_features,
+            out_features=orig_linear.out_features,
+            bias=orig_linear.bias is not None,
+            device=orig_linear.weight.device,
+            dtype=orig_linear.weight.dtype,
+        )
+        LinearAdapter._init_adapter(self, orig_linear)
 
-        self.orig_linear = orig_linear
-        self.dim = dim
-        self.scale = alpha / dim
+    @staticmethod
+    def _init_adapter(
+        obj,
+        orig_linear=None,
+        dim=8,
+        alpha=32,
+        dropout=0.1,
+        dropout_position='post',
+        lora_A_init_method='xavier',
+        lora_dtype=None,
+    ):
+        obj.dim = dim
+        obj.scale = alpha / dim
 
         # Freezer
-        device = self.orig_linear.weight.device
-        self.orig_linear.weight.requires_grad = False
-        if self.orig_linear.bias is not None:
-            self.orig_linear.bias.requires_grad = False
-
-        in_features = self.orig_linear.in_features
-        out_features = self.orig_linear.out_features
-        dtype = lora_dtype or self.orig_linear.weight.dtype
-
-        self.lora_a = nn.Parameter(torch.zeros((in_features, dim), dtype=dtype, device=device))
-        self.lora_b = nn.Parameter(torch.zeros((dim, out_features), dtype=dtype, device=device))
+        device = obj.weight.device
+        obj.weight.requires_grad = False
+        if obj.bias is not None:
+            obj.bias.requires_grad = False
+        # copy weights
+        if orig_linear is not None:
```

Comment:
In the FSDPv1 path, orig_linear has not been allocated yes (operates within a `device("meta")` context), therefore we cannot copy any weights. Instead this function monkey-patches the input obj (in this case an `nn.Linear`)